### ISSUE TITLE
is_overlap_all_atoms is slow

### ISSUE THREAD

**Issue by ExpHP** [atztogo/spglib#43]
`is_overlap_all_atoms` appears to do the same thing that phonopy used to do before I added [`compute_permutation`](https://github.com/atztogo/phonopy/pull/47).  For this reason, it's showing up at the top of my perf reports.

This kind of stinks.  I can try moving parts of `compute_permutation` over to spglib, but it will be ironic that spglib never actually looks at the permutation and instead only cares that it succeeded.

**Comment by atztogo** [comment]
In the case of phonopy, the symmetry is that already known. In the case of spglib, symmetry is that to be searched. So the permutation does not need to appear. But the approach to sort by distance is valid to improve spglib performance, I think. Another trick to employ the crystallographic constraints may exist, which I didn't carefully explored.

**Comment by ExpHP** [comment]
>  Another trick to employ the crystallographic constraints may exist, which I didn't carefully explored.

I'm not sure what you have in mind here, but this made me think of something:

Any property of the positions which:
* can be represented exactly as an integer or a sequence of integers
* does not change as a symmetry operation moves the atom to a new position

can be easily included in the sorting scheme.  Some examples I can think of are atom type, and the number of unique points in the position's symmetry star (if it is possible to compute that at this point in the code).  Basically, instead of sorting by lattice point distance, you sort lexicographically by `(property_A, property_B, ..., lattice_distance)`.

In addition to optimizing a search for *all* overlapping positions, this could also help provide an optimized test for checking if it is even **possible** for all positions to overlap.  Basically, you take the smallest group of atoms that have equal values for all of the integer properties, apply the rotation to one of them, and then check for an overlapping position in that small group.

...

But that's all something to think of *after* adding the most important optimizations.

**Comment by ExpHP** [comment]
Note: I'm working on a prototype implementation right now.  It includes atom types in the sorting scheme simply because it's really easy to do (the other optimizations I mentioned however are not worth the trouble  right now)

**Comment by atztogo** [comment]
Yes, I use atom type for selecting candidates of translation parts of space group operations for their search.

As an example of crystallographic constraints, in ``search_translation_part`` (https://github.com/atztogo/spglib/blob/master/src/symmetry.c#L530) for space group operation search, it can be escaped from the for loop when only one translation is found because I know that the cell given to this function is always a primitive cell at the current spglib, though this is not true in the very old versions. Here I have to mention that for the pure translation search (rigid translation), equivalently the case with ``is_identity=1``, multiple translations have to be searched.

As an example, I tested with inserting a break in the else statement. The spglib unit tests ran without any failure. Unfortunately the performance improvement was little for the unittest cases for which the small cells are used. The performance gain may be large for larger cells.

```c
      atoms_found[i] = 1;
      num_trans++;
      if (is_identity) {
        num_trans += search_pure_translations(atoms_found,
                                              cell,
                                              trans,
                                              symprec);
      } else {
        break;
      }
```

**Comment by ExpHP** [comment]
> it can be escaped from the for loop when only one translation is found because I know that the cell given to this function is always a primitive cell at the current spglib, though this is not true in the very old versions. Here I have to mention that for the pure translation search (rigid translation), equivalently the case with `is_identity=1`, multiple translations have to be searched.

I am confused.  Normally when I hear the phrase "primitive cell," I think of a cell that has no pure translational symmetries other than lattice points.  Such a cell should have exactly one possible translation for every point group operator, including the identity (in fact, the identity is easiest, because its translation must be zero!).  Is that not what you mean when you say this cell is "primitive"?

---

While I still don't have a complete picture myself of how spglib works, here is how I've always assumed the library might work:

1. Reduce the lattice matrix in some way (Niggli reduction?) that makes the `x - nint(x)` strategy valid for computing shortest distances.
2. Find pure translations.
3. Construct the primitive cell to eliminate these pure translations.  Reduce it like step 1.
4. Find the point group of the primitive lattice.
5. For each operator in that point group, search for translations that makes it a valid symmetry for the structure.  Each operator will either have a single translation, or it will have no translations (in which case it is simply not part of the structure's point group).  This is what I assumed `get_translation` was for.

Is this general outline correct or does it do something very different?

**Comment by ExpHP** [comment]
**Note:** After optimizing `is_overlap_all_atoms`, it looks like `search_pure_translations` is the next slow function, which is unfortunate because I am currently confused about what this function does.  (what could it possibly be computing that requires searching O(n^3) combinations of atoms?  It does not make sense to me...)

**Comment by atztogo** [comment]
> I am confused. Normally when I hear the phrase "primitive cell," I think of a cell that has no pure translational symmetries other than lattice points. Such a cell should have exactly one possible translation for every point group operator, including the identity (in fact, the identity is easiest, because its translation must be zero!). Is that not what you mean when you say this cell is "primitive"?

In the space group operation search, what we need for the computational application is usually the set of the coset representatives with respect to the translation group. So spglib does this job. A space group operation (or equivalently here a coset representative) is denoted by (R|t) where R is the rotation part represented by a 3x3 integer matrix and t is the translation part represented by a 3x1 (non-integer) row matrix (i.e. a vector). I meant about the translation part. For symmorphic space group, with appropriate choice of origin and setting, t of all space group operations modulo 1 can be chosen as (0,0,0)^T, but in the case of spglib, the origin is not known at this point, so anyway the translation part has to be searched.

---

For the algorithm of spglib, I am currently writing a manuscript. I will send a draft version soon if you like. I leave the flowchart that I draw for this manuscript. [flowchart.pdf](https://github.com/atztogo/spglib/files/1577687/flowchart.pdf)

What you wrote 1-6 are roughly correct. But it's not so easy, because we don't know origin and setting. At no.4, it is not the Bravais lattice, but it's the lattice for the primitive cell. If we can know Bravais lattice, it means we know already setting ('origin' and 'setting' are technical terms in crystallography). At no. 6, we will find the origin.

---

``search_pure_translations`` is indeed slow. For the case that there are many pure translations inside input cell, recently I implemented a trick to use group property of translation to improve the search speed and this is what you discuss. But this is written complicated, and for the case where pure translations are few in the input cell, we can expect to make it fact even improving its detailed algorithm. For further improvement, numerical technique is necessary, for which we should focus on ``search_translation_part``. I think your algorithm with sorting by distance will work. Here I think the distance is not necessarily to be the Euclidean distance.

---

> I believe this can be made considerably more efficient. Luckily for us, if we know the permutation that describes t_new, then the j loop becomes unnecessary.

I agree. But do we know the permutation?

**Comment by ExpHP** [comment]
> If we can know Bravais lattice, it means we know already setting ('origin' and 'setting' are technical terms in crystallography).

Oops, it looks like I may have been misusing the phrase "Bravais lattice."  To clarify, I did not mean classifications like "orthorhombic" and etc., but rather, I'm referring to the (infinitely large) mathematical set of 3-vectors that represent translational symmetry of a structure.  (or in terms of computation, I mean "a lattice matrix *up to* unimodular transforms")

I guess this is what most people in the field simply call a "lattice."

---

> > I believe this can be made considerably more efficient. Luckily for us, if we know the permutation that describes t_new, then the j loop becomes unnecessary.
>
> I agree. But do we know the permutation?

The code calls `is_total_overlap_all_atoms` before entering that function, and the fastest algorithm I know of for a function like `is_total_overlap_all_atoms` is basically identical to `compute_permutation`.  So in my local branch I've already made this optimization.

**However...** it looks like I was wrong about `search_pure_translations` being in need of optimization. (I just had a bug that was making it appear super slow).  After fixing that bug, I see that my optimizations to `search_pure_translations` have zero impact on overall performance for my current test case, so I guess it can be left alone for now.
